include/kernel.h

```diff
@@ -0,0 +1,531 @@
+ *     wakes up a thread (see #kevent). This is the most common scenario:
+ *     an event connected to an interrupt is fired, and a higher priority
+ *     thread waiting for the event is woken up.
+ * 
+ *   * Cooperatively: when a thread explicitly calls a blocking function
+ *     like #kthread_sleep or #kthread_wait_event to wait for an event.
+ *     The kernel will put the current thread to sleep and schedule the
+ *     highest-priority thread among the ready ones, including threads
+ *     that have a priority lower than the one that goes to sleep. When
+ *     the event is fired, the lower-priority thread will be preempted
+ *     to give back control to the thread that just woke up.
+ *
+ *   * Cooperatively: when a thread explicitly calls #kthread_yield to
+ *     leave the CPU to other threads that might be ready at the same time.
+ *     Notice that, given the hard real-time guarantee, this only makes
+ *     sense if there are multiple ready threads at the same priority level,
+ *     as calling #kthread_yield will not schedule a lower-priority thread.
+ *     It is normally not necessary to call #thread_yield.
+ *
+ * In addition to threads (#kthread_t) and events (#kevent_t), the kernel
+ * library offers a message passing communication primitive called "message
+ * box" (#kmbox_t). A mailbox is a circular buffer of messages called
+ * "mails"; any number of threads can try to send mails into a mailbox,
+ * and any number of threads can try to read mails from a mailbox. Once a
+ * thread receives a mail from a mailbox, the mail is removed and other
+ * threads will not see it anymore. 
+ *
+ * Mailboxes can be useful as communicative primitives among threads
+ * (eg: to send tasks to other threads), or a synchronization primitive
+ * to wait for multiple events (see #kmbox_attach_event).
+ *
+ * @{
+ */
+
+/**
+ * @brief a kernel thread for parallel execution
+ *
+ * This structure represents a thread that can be scheduled
+ * for parallel execution. Create a thread with #kthread_new,
+ * allocating the required memory for the stack.
+ *
+ * Normally, there's no need to explicitly manage the lifetime
+ * or state of a thread. Once the thread is created, it is
+ * immediately started, and the kernel will scheduled it when
+ * required. If the thread exits (by simply returning by its
+ * entry point function), it will be cleaned up and the memory
+ * released.
+ *
+ * #kthread_new returns a reference to the thread (a pointer
+ * to kthread_t) that can be used to externally manage the
+ * thread like changing its priority or killing it.
+ */
+typedef struct kthread_s
+{
+	/** Pointer to the top of the stack, which contains the thread register state */
+	reg_block_t* stack_state;
+	/** Current state of interrupt depth for this thread */
+	int interrupt_depth;
+	/** Size of the stack in bytes */
+	int stack_size;
+	/** Name of thread (for debugging purposes) */
+	char name[17];
```

Comment:
Nitpick, but would you feel comfortable having the `17` here `#define`d as a constant? This might help with copying from `strncpy`-style functions.

Reply:
Is it necessary for the name for be a char array at all rather than a pointer to a string? I expect run-time generated names to be pretty rare, and making the thread struct might be better for the cache, especially if the name string is stored in a rodata section further away  and never or rarely accessed.

Reply:
Good point. I was in general concerned about stack overlows corrupting `kthread_t` instances to the point of making debugging difficult so an array of chars survives better than a pointer, but I think this concern is now obsolete as I have simply moved the `kthread_t` instead after rather than before the stack. I'll change this to a pointer.